### PR TITLE
Basic session handling

### PR DESCRIPTION
**Pull request by ndoschek**
- Rebasing and minor adjustments
  - Update README
  - Cleanup imports
  - Fix spelling mistake in ModelServerPaths
  - Remove type specifiers on RHS
- Logging for WS and CLI help option
  - Add logging for websocket calls
  - Add usage message and help option to command line
- Extend session handling
  - Add subscribe, unsubscribe and broadcast methods
  - Wire session broadcasts with Http requests
  - Rename to JsonResponse for general reuse
  - Add data and confirm Json responses
  - Send all responses in same format
  - Use only simple class names for logging
  - Fix SchemaController status response
  - Update README
- Edit JUnitTest for SessionController

### CODE REVIEW

File: com.eclipsesource.modelserver.emf/src/main/java/com/eclipsesource/modelserver/emf/common/ModelController.java [hunk 1]
```diff
@@ -15,80 +15,110 @@
...
+				}
+		);
 	}
 
 	public Handler modelUrisHandler = ctx -> {
-		ctx.json(this.modelRepository.getAllModelUris());
+		ctx.json(JsonResponse.data(this.modelRepository.getAllModelUris()));
 	};
 
+	private Optional<EObject> getEObjectFromContext(Context ctx) {
```
Comment (edgarmueller) [inline com.eclipsesource.modelserver.emf/src/main/java/com/eclipsesource/modelserver/emf/common/ModelController.java]: Rename to `readEObjectFromContext` or just `readEObject`.

File: com.eclipsesource.modelserver.emf/src/main/java/com/eclipsesource/modelserver/emf/common/ModelController.java [hunk 2]
```diff
@@ -15,80 +15,110 @@
...
+		final EMFJsonConverter emfJsonConverter = new EMFJsonConverter();
+		Optional<EObject> eObject = Optional.empty();
```
Comment (edgarmueller) [inline com.eclipsesource.modelserver.emf/src/main/java/com/eclipsesource/modelserver/emf/common/ModelController.java]: Remove local var, use inline returns instead.

File: com.eclipsesource.modelserver.emf/src/main/java/com/eclipsesource/modelserver/emf/common/ModelController.java [hunk 3]
```diff
@@ -15,80 +15,110 @@
...
+		try {
+			JsonNode json = JavalinJackson.getObjectMapper().readTree(ctx.body());
+			if (json.get("data").toString().equals("{}")) {
```
Comment (edgarmueller) [inline com.eclipsesource.modelserver.emf/src/main/java/com/eclipsesource/modelserver/emf/common/ModelController.java]: Extract `json.get("data").toString()` to local var.

File: com.eclipsesource.modelserver.emf/src/main/java/com/eclipsesource/modelserver/emf/common/ModelController.java [hunk 4]
```diff
@@ -15,80 +15,110 @@
...
 
 	@Override
 	public void delete(Context ctx, String modeluri) {
-		this.modelRepository.removeModel(modeluri);
+		if (this.modelRepository.hasModel(modeluri)) {
+			this.modelRepository.removeModel(modeluri);
+			ctx.json(JsonResponse.confirm("Model '" + modeluri + "' successfully deleted"));
+			this.sessionController.modelDeleted(modeluri);
+		} else {
+			handleError(ctx, 404, "Model not found, cannot be deleted!");
```
Comment (edgarmueller) [inline com.eclipsesource.modelserver.emf/src/main/java/com/eclipsesource/modelserver/emf/common/ModelController.java]: Include `modeluri` in log message.

File: com.eclipsesource.modelserver.emf/src/main/java/com/eclipsesource/modelserver/emf/common/SessionController.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+	public void subscribe(WsContext ctx, String modeluri) {
+		modelUrisToClients.computeIfAbsent(modeluri, clients -> ConcurrentHashMap.newKeySet()).add(ctx);
+		ctx.send(JsonResponse.success(Json.object(Json.prop("sessionId", Json.text(ctx.getSessionId())))));
+	}
+
+	public void unsubscribe(WsContext ctx) {
+		modelUrisToClients.forEach((modeluri, clients) -> {
+			clients.remove(ctx);
+			if (clients.isEmpty())
```
Comment (edgarmueller) [inline com.eclipsesource.modelserver.emf/src/main/java/com/eclipsesource/modelserver/emf/common/SessionController.java]: This looks like it should throw an ConcurrentModificationException (it doesn't :), but makes me feel uncomfortable), Use curly braces for if statement

File: com.eclipsesource.modelserver.emf/src/main/java/com/eclipsesource/modelserver/emf/common/SessionController.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+		);
+	}
+
+	public void modelDeleted(String modeluri) {
+		broadcastModelUpdate(modeluri, null);
+	}
+
+	private void broadcastModelUpdate(String modeluri, @Nullable EObject updatedModel) {
+		if (modelUrisToClients.containsKey(modeluri))
+			modelUrisToClients.get(modeluri).stream().filter(ctx -> ctx.session.isOpen())
```
Comment (edgarmueller) [inline com.eclipsesource.modelserver.emf/src/main/java/com/eclipsesource/modelserver/emf/common/SessionController.java]: Extract filtering of open sessions to own private helper method.

File: com.eclipsesource.modelserver.emf/src/main/java/com/eclipsesource/modelserver/emf/common/SessionController.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+				.forEach(session -> {
+					session.send(JsonResponse.data(updatedModel));
+				});
+	}
+
+	private void broadcastError(String modeluri, String errorMessage) {
+		modelUrisToClients.get(modeluri).stream().filter(ctx -> ctx.session.isOpen())
+			.forEach(session -> {
```
Comment (edgarmueller) [inline com.eclipsesource.modelserver.emf/src/main/java/com/eclipsesource/modelserver/emf/common/SessionController.java]: No need for curly braces

File: com.eclipsesource.modelserver.emf/src/main/java/com/eclipsesource/modelserver/emf/common/SessionController.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+	}
+
+	public void modelDeleted(String modeluri) {
+		broadcastModelUpdate(modeluri, null);
+	}
+
+	private void broadcastModelUpdate(String modeluri, @Nullable EObject updatedModel) {
+		if (modelUrisToClients.containsKey(modeluri))
+			modelUrisToClients.get(modeluri).stream().filter(ctx -> ctx.session.isOpen())
+				.forEach(session -> {
```
Comment (edgarmueller) [inline com.eclipsesource.modelserver.emf/src/main/java/com/eclipsesource/modelserver/emf/common/SessionController.java]: No need for curly braces

File: com.eclipsesource.modelserver.emf/src/test/java/com/eclipsesource/modelserver/emf/common/SessionControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+		final WsContext client = mock(WsContext.class);
+		when(client.send(anyString())).thenReturn(CompletableFuture.completedFuture(null));
+		when(client.getSessionId()).thenReturn(UUID.randomUUID().toString());
+		when(client.pathParam("modeluri")).thenReturn("fancytesturi");
+		sessionController.subscribe(client, client.pathParam("modeluri"));
+
+		assertTrue(sessionController.isClientSubscribed(client));
+
+		sessionController.unsubscribe(client);
```
Comment (edgarmueller) [inline com.eclipsesource.modelserver.emf/src/test/java/com/eclipsesource/modelserver/emf/common/SessionControllerTest.java]: Extract unsubscribe assert into own test. Some more tests would be good (e.g. if `modelChanged` is called upon change/deletion), but let's add them via a separate PR.

**Review by edgarmueller — COMMENTED** [review]
Looks good! I've made some comments, maybe you can have a look.

File: com.eclipsesource.modelserver.emf/src/main/java/com/eclipsesource/modelserver/emf/common/SessionController.java (hunk 5 above)
Comment (ndoschek) [reply com.eclipsesource.modelserver.emf/src/main/java/com/eclipsesource/modelserver/emf/common/SessionController.java]: I used the `Iterator` interface now to remove the entry if no clients are left.

File: com.eclipsesource.modelserver.emf/src/test/java/com/eclipsesource/modelserver/emf/common/SessionControllerTest.java (hunk 9 above)
Comment (ndoschek) [reply com.eclipsesource.modelserver.emf/src/test/java/com/eclipsesource/modelserver/emf/common/SessionControllerTest.java]: Extracted the test for unsubscribe for now and added an issue for extending unit tests (#26)

**Review by edgarmueller — APPROVED** [review]

**Pull request merged**